flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ResourceManager.java

```diff
@@ -894,17 +900,21 @@ public void grantLeadership(final UUID newLeaderSessionID) {
 
 				// clear the state if we've been the leader before
 				if (getFencingToken() != null) {
-					clearState();
+					clearStateInternal();
 				}
 
 				setFencingToken(newResourceManagerId);
 
 				slotManager.start(getFencingToken(), getMainThreadExecutor(), new ResourceActionsImpl());
 
-				getRpcService().execute(
-					() ->
+				prepareLeadershipAsync()
+					.exceptionally(t -> {
+						onFatalError(t);
+						return null;
+					})
+					.thenRunAsync(() ->
 						// confirming the leader session ID might be blocking,
-						leaderElectionService.confirmLeaderSessionID(newLeaderSessionID));
+						leaderElectionService.confirmLeaderSessionID(newLeaderSessionID), getRpcService().getExecutor());
```

Comment:
Maybe we could refactor the `grantLeadership` the following way:
```
/**
	 * Callback method when current resourceManager is granted leadership.
	 *
	 * @param newLeaderSessionID unique leadershipID
	 */
	@Override
	public void grantLeadership(final UUID newLeaderSessionID) {
		final CompletableFuture<Boolean> acceptLeadershipFuture = CompletableFuture.supplyAsync(
			() -> tryAcceptLeadership(newLeaderSessionID),
			getUnfencedMainThreadExecutor()).thenCompose(Function.identity());

		final CompletableFuture<Void> confirmationFuture = acceptLeadershipFuture.thenAcceptAsync(
			(Boolean acceptLeadership) -> {
				if (acceptLeadership) {
					// confirming the leader session ID might be blocking,
					leaderElectionService.confirmLeaderSessionID(newLeaderSessionID);
				}
			},
			getRpcService().getExecutor());

		confirmationFuture.whenComplete(
			(Void ignored, Throwable throwable) -> {
				if (throwable != null) {
					onFatalError(ExceptionUtils.stripCompletionException(throwable));
				}
			});
	}

	private CompletableFuture<Boolean> tryAcceptLeadership(UUID newLeaderSessionID) {
		if (leaderElectionService.hasLeadership(newLeaderSessionID)) {
			final ResourceManagerId newResourceManagerId = ResourceManagerId.fromUuid(newLeaderSessionID);

			log.info("ResourceManager {} was granted leadership with fencing token {}", getAddress(), newResourceManagerId);

			// clear the state if we've been the leader before
			if (getFencingToken() != null) {
				clearStateInternal();
			}

			setFencingToken(newResourceManagerId);

			slotManager.start(getFencingToken(), getMainThreadExecutor(), new ResourceActionsImpl());

			return prepareLeadershipAsync().thenApply(ignored -> true);
		} else {
			return CompletableFuture.completedFuture(false);
		}
	}
```

Reply:
I don't understand why it's needed to check for leadership in `tryAcceptLeadership`. The zk `LeaderElectionService` already does that:
```
		if (leaderLatch.hasLeadership()) {
			// check if this is an old confirmation call
			synchronized (lock) {
				if (running) {
					if (leaderSessionID.equals(this.issuedLeaderSessionID)) {
						confirmedLeaderSessionID = leaderSessionID;
						writeLeaderInformation(confirmedLeaderSessionID);
					}
				} else {
					LOG.debug("Ignoring the leader session Id {} confirmation, since the " +
						"ZooKeeperLeaderElectionService has already been stopped.", leaderSessionID);
				}
			}
		} else {
			LOG.warn("The leader session ID {} was confirmed even though the " +
					"corresponding JobManager was not elected as the leader.", leaderSessionID);
		}
```

Reply:
The check would help to guard against a concurrent revoke leadership which is triggered just before the aynchronous grant leadership call is executed. That way we would not unnecessarily initialize internal components which would be stopped by the revoke leadership async callback. At the moment this is not very likely to happen but if we make `clearState` asynchronous and wait in the `grantLeadership` method to complete the cleanup before calling the leadership callback, it gets more and more likely.